File: content/problems/0496-next-greater-element-i/solutions/monotonic-decreasing-stack/Solution.kt
class Solution {
    fun nextGreaterElement(nums1:IntArray,nums2:IntArray):IntArray{
        val next=hashMapOf<Int,Int>();
        val stack=ArrayDeque<Int>();
        for(value in nums2){
            while(stack.isNotEmpty()&&stack.last()<value)next[stack.removeLast()]=value;
            stack.addLast(value)
        };
        return IntArray(nums1.size){
            next[nums1[it]]?:-1
        }
    }
}

File: content/problems/0496-next-greater-element-i/solutions/monotonic-decreasing-stack/solution.cpp
class Solution {
    public: vector<int> nextGreaterElement(vector<int>& nums1,vector<int>& nums2){
        unordered_map<int,int> next;
        vector<int> stack;
        for(int value:nums2){
            while(!stack.empty()&&stack.back()<value){
                next[stack.back()]=value;
                stack.pop_back();
            }
            stack.push_back(value);
        }
        vector<int> answer;
        for(int value:nums1)answer.push_back(next.count(value)?next[value]:-1);
        return answer;
    }
};
